src/projection.jl

```diff
@@ -31,114 +68,269 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
     end
 end
 
-# fallback (structs)
-function ProjectTo(x::T) where {T}
-    # Generic fallback for structs, recursively make `ProjectTo`s all their fields
+# Structs
+function generic_projectto(x::T; kw...) where {T}
+    # Generic fallback, recursively make `ProjectTo`s for all their fields
     fields_nt::NamedTuple = backing(x)
-    return ProjectTo{T}(map(ProjectTo, fields_nt))
+    fields_proj = map(fields_nt) do x1
+        if x1 isa Number || x1 isa AbstractArray
+            ProjectTo(x1)
+        else
+            x1
+        end
+    end        
+    # We can't use `T` because if we have `Foo{Matrix{E}}` it should be allowed to make a
+    # `Foo{Diagaonal{E}}` etc. We assume it has a default constructor that has all fields 
+    # but if it doesn't `construct` will give a good error message.
+    wrapT = T.name.wrapper
+    return ProjectTo{wrapT}(; fields_proj..., kw...)
 end
 function (project::ProjectTo{T})(dx::Tangent) where {T}
     sub_projects = backing(project)
     sub_dxs = backing(canonicalize(dx))
-    _call(f, x) = f(x)
-    return construct(T, map(_call, sub_projects, sub_dxs))
-end
-
-# should not work for Tuples and NamedTuples, as not valid tangent types
-function ProjectTo(x::T) where {T<:Union{<:Tuple,NamedTuple}}
-    return throw(
-        ArgumentError("The `x` in `ProjectTo(x)` must be a valid differential, not $x")
-    )
+    maybe_call(f::ProjectTo, x) = f(x)
+    maybe_call(f, x) = f
+    return construct(T, map(maybe_call, sub_projects, sub_dxs))
 end
 
 # Generic
+(::ProjectTo{T})(dx::T) where {T} = dx 
+(::ProjectTo{T})(dx::AbstractZero) where {T} = dx
+(::ProjectTo{T})(dx::NotImplemented) where {T} = dx
+
+ProjectTo() = ProjectTo{Any}()  # trivial case, exists so that maybe_call(f, x) knows what to do
+(x::ProjectTo{Any})(dx) = dx
+
+# Thunks
+(project::ProjectTo)(dx::Thunk) = Thunk(project ∘ dx.f)
+(project::ProjectTo)(dx::InplaceableThunk) = project(dx.val)  # can't update in-place part
 (project::ProjectTo)(dx::AbstractThunk) = project(unthunk(dx))
```

Comment:
Don't need this one do we?
```suggestion
```

Reply:
Throughout, I want the list of things it's applicable without error to be simple. If it does Ref{Float64} is must also  not error on Ref{Any}, even if it can't do much. This is why I was thinking that, if it accepts some thunks, it should accept all of them.

Reply:
I see.
I think that is a secondary concern to always doing something correct.

This in particular right now is a problem beyond that though.
It is dead code.
This method can't be hit, and can't be tested.
It makes the code-base harder to understand, or change.

e.g.  it also needs to be removed if we remove the one for InplacableThunk.

Reply:
To break this you'd need an AbstractThunk which doesn't define `thunk`, and then I think you've got bigger problems.

Reply:
Exact;y why this is currently dead code that can't be hit.
Dead code should not remain in the package.

Reply:
Now there's one tested method, which prints a bright red warning message.